compiler-core/src/error.rs

```diff
@@ -2927,10 +2928,79 @@ See: https://tour.gleam.run/advanced-features/use/");
                         }),
                     }
                 },
+                TypeError::BadName { location, name } => {
+                    let label = "This is not a valid name";
+
+                let text = wrap_format!("Hint: Names start with a lowercase letter and contain a-z, 0-9, or _.
+Try: {}", name.to_snake_case());
+
+                Diagnostic {
+                    title: "Invalid name".into(),
+                    text,
+                    hint: None,
+                    level: Level::Error,
+                    location: Some(Location {
+                        label: Label {
+                            text: Some(label.to_string()),
+                            span: *location,
+                        },
+                        path: path.clone(),
+                        src: src.clone(),
+                        extra_labels: vec![],
+                    }),
+                }
+                },
+
+                TypeError::BadDiscardName { location, name } => {
+                    let label = "This is not a valid discard name";
+
+                let text = wrap_format!("Hint: Discard names start with _ and contain a-z, 0-9, or _.
+Try: _{}", name.to_snake_case());
+
+                Diagnostic {
+                    title: "Invalid discard name".into(),
+                    text,
+                    hint: None,
+                    level: Level::Error,
+                    location: Some(Location {
+                        label: Label {
+                            text: Some(label.to_string()),
+                            span: *location,
+                        },
+                        path: path.clone(),
+                        src: src.clone(),
+                        extra_labels: vec![],
+                    }),
+                }
+                },
+
+                TypeError::BadUpName { location, name } => {
+                    let label = "This is not a valid upname";
```

Comment:
"upname" doesn't mean anything to a Gleam programmer, it's an internal name to the compiler and could change at any time. 

Let's merge these 3 new error variants into one `BadName` variant and have an enum to say what the context is:

```rust
pub enum BadNameKind {
    Type,
    TypeVariable,
    CustomTypeVariant,
    Variable,
    Function,
}
```

Reply:
> "upname" doesn't mean anything to a Gleam programmer, it's an internal name to the compiler and could change at any time.

Right, I figured that but I stuck with "upname" since that was used in error messages previously